### PR TITLE
[css-conditional-5] Draft spec for scroll-state() #6402

### PR DESCRIPTION
**Pull request by lilles**
Per resolution in [1], add scroll-state() as a query function to query sticky, snap, and overflow scroll states with a new container-type for scroll-state queries.

[1] https://github.com/w3c/csswg-drafts/issues/6402#issuecomment-1812973013

### CODE REVIEW

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1115]: I find this a bit confusing.  Is the "current state" mentioned in the first sentence the same as the "scroll state" in the second sentence?

In the second sentence, avoid the passive voice by saying "update the scroll state" rather than "the scroll state ... is updated".

It's also not clear to me how this prevents cycles.  It sounds like this is saying to re-run an arbitrary number of times.  Or does it mean to re-run just once and then stop?

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1132]: I assume that if the query container is not `position: sticky` then none of the values match?  It would be good to state that explicitly.  (It might make sense to have a bit less formal definition for each value in the `<dl>` since they have a lot in common.)

I also think that "visually shifted" should perhaps link to the definition in css-position-3.

It might be worth noting that a container can be stuck to two opposite sides at the same time.  (At least, after reading the text in css-position-3, I think that's the case.)

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1180]: I think "the container" is "the query container", right?

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1198]: I don't think "Sticky positioning" is the right header here.

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1150]: link the logical values to their [definitions](https://drafts.csswg.org/css-writing-modes-4/#logical-directions)?

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1230]: link the logical values to their [definitions](https://drafts.csswg.org/css-writing-modes-4/#logical-directions)?

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1208]: I think it should be clear that none of these match if the query container isn't a scroll container.

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1217]: I think it should be clear if these match when there is overflow even if the container is not scrollable in that direction.

I also wonder if there's demand for both variants, that is, an `(overflowing:left)` that is true only for RTL or vertical-rl pages where you can actually scroll to the left, and a differently named `(overflowing: left)` that is true if there is overflow to the left whether or not it can be scrolled to.

(I don't think it's worth distinguishing based on `overflow: hidden`... particularly since that's *programatically* scrollable.  But if things do depend on "is it scrollable in that direction" it should be clear whether or not `hidden` makes a difference.)

**Review by dbaron — COMMENTED** [review]
This looks reasonable.  One big question at the top (about whether it's supposed to be in this file at all), and then a bunch of specific comments on the text.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:463]: > One high-level question here is whether these edits should go in css-conditional-5. The resolution explicitly called for a delta spec. (I realized the base spec has moved from css-contain-3 to css-conditional-5, but I think not doing a delta spec is still surprising.)
> 
> If this gets turned into a delta spec... then some of my later comments probably don't make as much sense.

I was unsure and asked Tab, who proposed adding it to this level (but he's not editing this spec).

Comment (lilles) [reply css-conditional-5/Overview.bs:463]: > > One high-level question here is whether these edits should go in css-conditional-5. The resolution explicitly called for a delta spec. (I realized the base spec has moved from css-contain-3 to css-conditional-5, but I think not doing a delta spec is still surprising.)
> > If this gets turned into a delta spec... then some of my later comments probably don't make as much sense.
> 
> I was unsure and asked Tab, who proposed adding it to this level (but he's not editing this spec).

Adding @fantasai for another opinion. Should scroll-state() queries be added to a css-conditional-6 delta spec?

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:1153]: > I assume that if the query container is not `position: sticky` then none of the values match? It would be good to state that explicitly. (It might make sense to have a bit less formal definition for each value in the `<dl>` since they have a lot in common.)

Done.

> I also think that "visually shifted" should perhaps link to the definition in css-position-3.

"visually shifted" does not have a dfn in that spec.

> It might be worth noting that a container can be stuck to two opposite sides at the same time. (At least, after reading the text in css-position-3, I think that's the case.)

Done.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:1242]: > I think it should be clear if these match when there is overflow even if the container is not scrollable in that direction.
> 
> I also wonder if there's demand for both variants, that is, an `(overflowing:left)` that is true only for RTL or vertical-rl pages where you can actually scroll to the left, and a differently named `(overflowing: left)` that is true if there is overflow to the left whether or not it can be scrolled to.

The intention is that matching is based on the current scroll position, so `left` and `right` will both match when a horizontal scrollbar is not a 0 nor at the max position. If there is content overflowing to the left of an ltr container, the content is not reachable through user scrolling, and `(overflowing:left)` will not match.

> (I don't think it's worth distinguishing based on `overflow: hidden`... particularly since that's _programatically_ scrollable. But if things do depend on "is it scrollable in that direction" it should be clear whether or not `hidden` makes a difference.)

I used the term [_scrollable overflow_](https://drafts.csswg.org/css-overflow-3/#scrollable-overflow) to exclude `overflow:hidden` and overflow that's not reachable through user scrolling. I've made it more explicit in the text now.

File: css-conditional-5/Overview.bs
Comment (svgeesus) [reply css-conditional-5/Overview.bs:463]: I think it is fine to put this in Conditional 5, which is already a delta over Conditional 3 and 4.

**Review by svgeesus — APPROVED** [review]
Overall approve, as I prefer to see good PR merged sooner and details worked out later. But I will wait for one of the other editors to do the merge. Not all of the @dbaron comments have been resolved.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs]: > I find this a bit confusing. Is the "current state" mentioned in the first sentence the same as the "scroll state" in the second sentence?

Changed the latter "scroll state" to "current state" which should be clearer.

> In the second sentence, avoid the passive voice by saying "update the scroll state" rather than "the scroll state ... is updated".

Done.

> It's also not clear to me how this prevents cycles. It sounds like this is saying to re-run an arbitrary number of times. Or does it mean to re-run just once and then stop?

Just a single time. Clarified.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:479]: Moved the definition now.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs]: > This feels like it's trying to be a definition of what the query container is in this case, but it doesn't feel like it gives a clear definition. I think it either (a) needs to be a clear definition or (b) needs to sound less like it's trying to define what the query container is. (I think (b) is the right thing because it's defined above, although I'm not sure.)

I changed the text to focus on scroll-state queries instead of the query container.

**Comment by lilles** [comment]
> Overall approve, as I prefer to see good PR merged sooner and details worked out later. But I will wait for one of the other editors to do the merge. Not all of the @dbaron comments have been resolved.

Thanks!

**Comment by svgeesus** [comment]
@dbaron are you now satisfied with the responses to your helpful comments on this PR? If so please r+

File: css-conditional-5/Overview.bs
Comment (dbaron) [reply css-conditional-5/Overview.bs:1154]: It's still not clear what happens when the sticky algorithm shifts both sides of the container in the same axis.  It's OK to address that as a followup, though.

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1231]: Hmmm, now that I'm rereading this text and your replies... what does "current scroll progress" mean here.  Does it mean that an LTR overflow:scroll container is `overflowing(left)` once it's been scrolled away from its initial horizontal position?  (That contradicts the following sentence, which I had suggested you add!)

File: css-conditional-5/Overview.bs
Comment (dbaron) [inline css-conditional-5/Overview.bs:1121]: At some point it needs to be clear what order this "single time" update has relative to the other features that do a "single time" update (since when both features are used together it can affect the results).  That doesn't need to be resolved right now, but it might be worth adding an issue.

Comment (dbaron) [reply css-conditional-5/Overview.bs:1121]: (or maybe the "scroll steps" don't have other things that do single-time update?  There certainly are a bunch of other features that work that way, but maybe none that hook into scroll steps?)

File: css-conditional-5/Overview.bs
Comment (dbaron) [reply css-conditional-5/Overview.bs:689]: I think this still needs a grammar production, not just prose.

**Review by dbaron — APPROVED** [review]
I left four new comments (two of which show up directly above, and two of which only show up in the middle of the earlier threads), but I'm fine with this landing either in its current state or with an attempt to address those.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:1121]: > At some point it needs to be clear what order this "single time" update has relative to the other features that do a "single time" update (since when both features are used together it can affect the results). That doesn't need to be resolved right now, but it might be worth adding an issue.

That is what issue #10796 mentioned at the top of this section is about.

> (or maybe the "scroll steps" don't have other things that do single-time update? There certainly are a bunch of other features that work that way, but maybe none that hook into scroll steps?)

The "scroll steps" updating of the current state takes effect in the "update style and layout" in the resizeObserver loop in the html spec. The state update with an extra style and layout update _after_ the resizeObserver loop does have the issue of which does that first when taking anchor positioning and scroll timelines into account. That's what issue #10796 tries to address.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:689]: It refers to the `media feature` syntax just like the size and style features do, so if that needs further fixing it should be done consistently across the feature types.

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs:1154]: > It's still not clear what happens when the sticky algorithm shifts both sides of the container in the same axis. It's OK to address that as a followup, though.

How is that even possible? It can only be shifted in one direction per axis, right?

File: css-conditional-5/Overview.bs
Comment (lilles) [reply css-conditional-5/Overview.bs]: Reworded.

File: css-conditional-5/Overview.bs
Comment (dbaron) [reply css-conditional-5/Overview.bs:1154]: Ah, never mind.  I guess I missed a sentence earlier in the [spec's text](https://drafts.csswg.org/css-position-3/#stickypos-insets):

> If this results in a [sticky view rectangle](https://drafts.csswg.org/css-position-3/#sticky-view-rectangle) size in any axis less than the size of the [border box](https://drafts.csswg.org/css-box-4/#border-box) of the sticky box in that axis, then the effective [end](https://drafts.csswg.org/css-writing-modes-4/#css-end)-edge inset in the affected axis is reduced (possibly becoming negative) to bring the sticky view rectangle’s size up to the size of the border box in that axis (where end is interpreted relative to the [writing mode](https://drafts.csswg.org/css-writing-modes-4/#writing-mode) of the [containing block](https://drafts.csswg.org/css-display-4/#containing-block)).

that removes the possibility that this can happen.  (But otherwise, reading just the algorithm in the following paragraph, it seems like it could... although I admit it's not clear what the priority would be!)

**Pull request merged**
